Use improved /messages/flags API to mark messages read / unread.

amanagr opened the pull request:
UI extension of #23512.

<img width="1066" alt="image" src="https://user-images.githubusercontent.com/25124304/201951467-68883607-7e0f-4eb8-b54e-cbe2d3bb31dc.png">
<img width="1149" alt="image" src="https://user-images.githubusercontent.com/25124304/201951507-762359fc-0127-4a3a-98eb-617718abafd6.png">
<img width="1143" alt="image" src="https://user-images.githubusercontent.com/25124304/201951528-12e6a0d6-3989-4ec6-9948-b8f25423c257.png">

alya commented:
Hm, how do we have a percentage we can show? In the #design discussion, Tim mentioned that we won't know how many total messages will be marked as unread. Do we have that data after all?

timabbott commented on static/js/unread_ops.js:
```diff
@@ -13,36 +13,69 @@ import * as notifications from "./notifications";
+                    percentage_updated = (percentage_updated * 100).toFixed(2);
+                    ui_report.loading(
+                        $t_html(
+                            {defaultMessage: "Marking messages as read...{percentage_updated}%"},
```
We should use a unicode `...` rather than 3 dot characters if we're doing that. 

The other concern with doing a percentage is that the percentage is only accurate if !old_unreads_missing; so we will need an alternative string to display in the case that we don't know how many unreads there are (but do know there were at least ~50k to start with).

alya replied:
Let's change the string to:

"Marking all messages as read… {percentage_updated}% complete."

alya replied:
In the >50K case, we can use:

"Marking all messages as read… N messages processed."

amanagr replied:
Ok, made those changes!

timabbott commented on static/js/unread_ops.js:
```diff
@@ -13,36 +13,69 @@ import * as notifications from "./notifications";
-    unread.declare_bankruptcy();
-    unread_ui.update_unread_counts();
+let unread_count_before_marking_all_msgs_read = 0;
+let messages_read_till_now = 0;
```
Oh, this is using recursion on the whole function, not a smaller helper inside; so the above suggestion is wrong.

amanagr replied:
Ok, fixed by moving these variables inside the function.

timabbott commented on static/js/unread_ops.js:
```diff
@@ -13,36 +13,69 @@ import * as notifications from "./notifications";
+        url: "/json/messages/flags",
+        data: opts,
+        success(data) {
+            if (!data.found_newest) {
```
Maybe add a block comment here explaining that we're only displaying a progress indicator in the case that there's so many messages to modify that we were not able to modify them all in a single request; if we can process them all in a single request, we expect the user to get immediate feedback in the form of the unread count disappearing.

amanagr replied:
Done!

timabbott commented on static/js/unread_ops.js:
```diff
@@ -13,36 +13,69 @@ import * as notifications from "./notifications";
+                    $t_html({defaultMessage: "Successfully marked all messages as read!"}),
+                    $("#request-progress-status-banner"),
+                    true,
+                );
```
There's a final detail here: If `old_unreads_missing`, then we may need to reload the browser afterwards, essentially in order to clear that state; but I think that will become the only situation where a reload might still need to exist in this code path. 

I think it's POSSIBLE that just setting `old_unreads_missing` to false would be correct behavior after this flow; but I'm not sure, and it's pretty cheap to just do a reload after someone marked 50k+ messages as read, so probably best to just do that.

amanagr replied:
Okay, added a `TODO` comment for it too.

timabbott commented:
> Hm, how do we have a percentage we can show? In the #design discussion, Tim mentioned that we won't know how many total messages will be marked as unread. Do we have that data after all?

This pull request changes both "mark all as read" and "mark as unread" to use the new API; in the former case, we USUALLY will know how many total unread messages there are to process. The exception is when there are more than 50K unreads, as in that case, the server only sends us the most recent 50k unreads; so we probably need a different string for that (rare but real) scenario to avoid displaying "123%" complete sometimes.

timabbott commented:
This looks pretty promising! How are you testing the corner cases -- do you just set the batch size to like 3?

amanagr commented:
> This looks pretty promising! How are you testing the corner cases -- do you just set the batch size to like 3?
Yes.

timabbott commented:
We should have the commit message for the mark-as-read part of this close https://github.com/zulip/zulip/issues/23406 and #14053.

timabbott commented on static/js/unread_ops.js:
```diff
@@ -9,32 +13,69 @@ import * as notifications from "./notifications";
-    unread.declare_bankruptcy();
-    unread_ui.update_unread_counts();
+// anchor message is not included in this calculation
+const NUM_OF_MESSAGES_UPDATED_PER_BATCH = 5000;
```
@alya this is the value to change to 1 to see the progress experience.

alya commented:
Thanks!

Playing with the PR, the confirmation banner disappears too quickly. We should keep it around long enough for the user to have a chance to read it.

alya commented:
I see the confirmation banner for marking messages as read when I do so in "All messages", but not when doing so for a stream or topic, which feels inconsistent. Does it depend on the number of messages affected?

alya commented:
After changing 5000 to 1, I got this error:
![Screen Shot 2022-11-15 at 1 25 13 PM](https://user-images.githubusercontent.com/2090066/202028514-fabbe63f-62f2-4765-b93c-f849f7bcc7c2.png)

alya commented:
Chatting with @timabbott , the success banner should be shown if, and only if, we showed a progress banner (same for "mark as read" and "mark as unread".

alya commented:
There seems to be something wrong with the % counter; I'm seeing it going about 100%.

alya commented on static/js/unread_ops.js:
```diff
@@ -9,32 +13,69 @@ import * as notifications from "./notifications";
+                } else {
+                    ui_report.loading(
+                        $t_html({defaultMessage: "Marking messages as read..."}),
+                        $("#request-progress-status-banner"),
```
I think this one won't be shown, since we won't have a progress indicator in this situation?

amanagr replied:
Ok!

alya commented on static/js/unread_ops.js:
```diff
@@ -9,32 +13,69 @@ import * as notifications from "./notifications";
+                loading_indicator_displayed = false;
+                ui_report.loading(
+                    $t_html({defaultMessage: "Successfully marked all messages as read!"}),
+                    $("#request-progress-status-banner"),
```
For the string, let's just use:

> Success!

We'll only show it after the progress banner, so it should be pretty clear what it's referring to.

alya commented:
For the colors, let's make the banners blue (like the upload banners, I guess?) throughout -- both in-progress and success banners, to avoid it being jarring. (If blue is awkward for some reason, green might be OK too.)

Please test in both light and dark theme; dark theme looks broken in the current PR:

![Screen Shot 2022-11-15 at 1 47 11 PM](https://user-images.githubusercontent.com/2090066/202032366-0caf6c10-2043-49d5-b45e-77de176a9b4e.png)

alya commented on static/js/unread_ops.js:
```diff
@@ -47,17 +88,44 @@ function process_newly_read_message(message, options) {
+                if (!loading_indicator_displayed) {
+                    ui_report.loading(
+                        $t_html({defaultMessage: "Marking messages as unread..."}),
+                        $("#request-progress-status-banner"),
```
Let's use:

> "Working... N messages marked as unread so far."

alya commented on static/js/unread_ops.js:
```diff
@@ -47,17 +88,44 @@ function process_newly_read_message(message, options) {
+                loading_indicator_displayed = false;
+                ui_report.loading(
+                    $t_html({defaultMessage: "Successfully marked requested messages as unread!"}),
+                    $("#request-progress-status-banner"),
```
Let's just use "Success!" here as well.

timabbott commented on static/js/unread_ops.js:
```diff
@@ -9,32 +13,69 @@ import * as notifications from "./notifications";
+        unread_count_before_marking_all_msgs_read = unread.get_unread_message_count();
+    }
+    const opts = {
+        anchor: "first_unread",
```
I think there was a thought that the recursive calls could use the anchor returned by the previous API response; this more obviously terminates and also is likely more efficient.

amanagr replied:
Yeah, added that.

amanagr commented on static/js/unread_ops.js:
```diff
@@ -13,36 +13,69 @@ import * as notifications from "./notifications";
 let loading_indicator_displayed = false;
-export function mark_all_as_read() {
-    unread.declare_bankruptcy();
-    unread_ui.update_unread_counts();
```
`declare_bankruptcy` is still used in tests, so left it out.

timabbott replied:
OK; that makes sense. Maybe worth adding a comment that this should only be called from tests.

amanagr commented:
> Playing with the PR, the confirmation banner disappears too quickly. We should keep it around long enough for the user to have a chance to read it.

Increased duration to 2.5s from 1s.

amanagr commented:
> https://github.com/zulip/zulip/pull/23579#issuecomment-1315883224

This is happening at the same time when we hit the API limit. I added `error` functions the server requests to better handles that case. Also, added a commit to handle the case of missing `message_details`.

amanagr commented:
> https://github.com/zulip/zulip/pull/23579#issuecomment-1315905289

What do you think of these colors to match the dark theme dropdown colors?
<img width="971" alt="image" src="https://user-images.githubusercontent.com/25124304/202151205-115af23e-2279-4d2e-b5da-55ff31ac33fb.png">
<img width="968" alt="image" src="https://user-images.githubusercontent.com/25124304/202151267-f72a00ef-7237-4b81-bc7f-23b178e2751b.png">

amanagr commented:
> There seems to be something wrong with the % counter; I'm seeing it going above 100%.

For some reason, this is still happening, I will try to take another look after some time; maybe something will popout.

amanagr commented:
> For some reason, this is still happening, I will try to take another look after some time; maybe something will popout.

My conclusion right now is that this is due to `unread.get_unread_message_count()` being incorrect after running `mark_as_unread_from_here`. Its value is lower than it should be. Both `mark_as_unread_from_here` and `mark_all_as_read` are correct after a reload.

amanagr commented on static/js/unread_ops.js:
```diff
@@ -185,14 +185,20 @@ export function process_unread_messages_event({message_ids, message_details}) {
     for (const message_id of message_ids) {
+        const message_info = message_details[message_id];
+        if (!message_info) {
+            // Server sent us the `message_id` but didn't sent us the `message_details` for
+            // it. This case is easy to reproduce after hitting an API limit when running `mark_as_unread_from_here`.
+            // TODO: This is most likely a server side bug and needs to be fixed there.
+            continue;
+        }
+
         const message = message_store.get(message_id);
 
         if (message) {
             message.unread = true;
         }
 
-        const message_info = message_details[message_id];
-
         let user_ids_string;
```
Oh! We actually want to process the `message_id` here, regardless of if the server sent us the `message_details` for it.

timabbott commented on static/js/unread_ops.js:
```diff
@@ -47,17 +55,84 @@ function process_newly_read_message(message, options) {
+                        ),
+                    xhr.responseJSON["retry-after"],
+                );
+            }
```
I think we need an else clause to show some sort of failure message.

amanagr replied:
Ok, added a blueslip error.

amanagr commented:
<img width="745" alt="image" src="https://user-images.githubusercontent.com/25124304/202286342-fbd0da77-17ef-4032-b56d-2a69cef66628.png">

updated mark as read text

alya commented:
Hm, I'm seeing a final count that's smaller than the max "processed N messages" count, which is confusing. The final count should refer to the same data, and should never be smaller.

amanagr commented:
Updated to remove `exit` from loading banners and reverted an accidental change of `const NUM_OF_MESSAGES_UPDATED_PER_BATCH = 10;`

amanagr commented on static/js/unread_ops.js:
```diff
@@ -14,36 +14,98 @@ import * as notifications from "./notifications";
+                    messages_read_till_now,
+                });
+            } else {
+                if (loading_indicator_displayed) {
+                    // Only show the success message if a progress banner was displayed.
+                    ui_report.loading(
```
The comment here was incomplete, so changed it.

timabbott replied:
Thanks!

timabbott commented:
Pushed back with these changes; one to fix the fact that we starting marking-as-unread in a different narrow if you clicked to another view while it was running, and another to fix the weird `message_details` bug.

```
diff --git a/static/js/unread_ops.js b/static/js/unread_ops.js
index 9d10b9995b..6cbda1d3e4 100644
--- a/static/js/unread_ops.js
+++ b/static/js/unread_ops.js
@@ -122,14 +122,18 @@ export function mark_as_unread_from_here(
     message_id,
     include_message = true,
     messages_marked_unread_till_now = 0,
+    narrow,
 ) {
+    if (narrow === undefined) {
+        narrow = JSON.stringify(message_lists.current.data.filter.operators());
+    }
     message_lists.current.prevent_reading();
     const opts = {
         anchor: message_id,
         include_anchor: include_message,
         num_before: 0,
         num_after: NUM_OF_MESSAGES_UPDATED_PER_BATCH,
-        narrow: JSON.stringify(message_lists.current.data.filter.operators()),
+        narrow,
         op: "remove",
         flag: "read",
     };
@@ -162,6 +166,7 @@ export function mark_as_unread_from_here(
                     data.last_processed_id,
                     false,
                     messages_marked_unread_till_now,
+                    narrow,
                 );
             } else if (loading_indicator_displayed) {
                 // If we were showing a loading indicator, then
@@ -192,6 +197,7 @@ export function mark_as_unread_from_here(
                             message_id,
                             false,
                             messages_marked_unread_till_now,
+                            narrow,
                         ),
                     xhr.responseJSON["retry-after"],
                 );
diff --git a/zerver/tornado/event_queue.py b/zerver/tornado/event_queue.py
index fa57b08b63..648fafc851 100644
--- a/zerver/tornado/event_queue.py
+++ b/zerver/tornado/event_queue.py
@@ -350,10 +350,25 @@ class EventQueue:
         event["id"] = self.next_event_id
         self.next_event_id += 1
         full_event_type = compute_full_event_type(event)
-        if full_event_type == "restart" or full_event_type.startswith("flags/"):
+        if full_event_type == "restart" or full_event_type.startswith("flags/add/"):
+            # virtual_events are an optimization that allows certain
+            # simple events, such as update_message_flags events that
+            # simply contain a list of message IDs to operate on, to
+            # be compressed together. This is primarily useful for
+            # flags/add/read, where normal Zulip usage will result in
+            # many small flags/add/read events as users scroll.
+            #
+            # We need to exclude flags/remove/read, because it has an
+            # extra message_details field that cannot be compressed.
+            #
+            # BUG: This compression algorithm is incorrect in the
+            # presence of mark-as-unread, since it does not respect
+            # the ordering of "mark as read" and "mark as unread"
+            # updates for a given message.
             if full_event_type not in self.virtual_events:
                 self.virtual_events[full_event_type] = copy.deepcopy(event)
                 return
+
             # Update the virtual event with the values from the event
             virtual_event = self.virtual_events[full_event_type]
             virtual_event["id"] = event["id"]
```

amanagr commented:
Taking care of the tests.

timabbott commented:
Opened https://github.com/zulip/zulip/issues/23600 for one follow-up.
